File: iOS-ObjC/MSALiOSobjC/ViewController.m
//------------------------------------------------------------------------------
//
// Copyright (c) Microsoft Corporation.
// All rights reserved.
//
// This code is licensed under the MIT License.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files(the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions :
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//
//------------------------------------------------------------------------------


#import "ViewController.h"
#import <MSAL/MSAL.h>

/// 😃 A View Controller that will respond to the events of the Storyboard.

@interface ViewController ()
    
    @end

@implementation ViewController
    
    MSALPublicClientApplication *application;
    
    // Update the below to your client ID you received in the portal. The below is for running the demo only
    NSString *clientId = @"db9191e7-9fa9-41e5-84e6-223298a0e9b2";
    // Additional variables for Auth and Graph API
    NSString *redirectUri = @"msaldb9191e7-9fa9-41e5-84e6-223298a0e9b2://auth";
    NSString *authorityString = @"https://login.microsoftonline.com/common";
    NSString *graphURI = @"https://graph.microsoft.com/v1.0/me/";
    
    /**
     Setup public client application in viewDidLoad
     */
    
    - (void)viewDidLoad {
        [super viewDidLoad];
        // Do any additional setup after loading the view.
        [self initMSAL];
    }
    
    // MARK: Initialization

    -(void) initMSAL{
        
        /**
         
         Initialize a MSALPublicClientApplication with a given clientID and authority
         
         - clientId:            The clientID of your application, you should get this from the app portal.
         - redirectUri:         A redirect URI of your application, you should get this from the app portal.
         If nil, MSAL will create one by default. i.e./ msauth.<bundleID>://auth
         - authority:           A URL indicating a directory that MSAL can use to obtain tokens. In Azure AD
         it is of the form https://<instance/<tenant>, where <instance> is the
         directory host (e.g. https://login.microsoftonline.com) and <tenant> is a
         identifier within the directory itself (e.g. a domain associated to the
         tenant, such as contoso.onmicrosoft.com, or the GUID representing the
         TenantID property of the directory)
         - error                The error that occurred creating the application object, if any, if you're
         not interested in the specific error pass in nil.
         */
        
        NSError *error = nil;
        
        NSURL *authorityURL = [NSURL URLWithString:(authorityString)];
        
        MSALAuthority *authority = [MSALAuthority authorityWithURL:authorityURL error:nil];
        
        
        MSALPublicClientApplicationConfig *pcaConfig = [[MSALPublicClientApplicationConfig alloc] initWithClientId:clientId
                                                                                                       redirectUri:redirectUri
                                                                                                         authority:authority];
        
        application = [[MSALPublicClientApplication alloc] initWithConfiguration:pcaConfig error:&error];
        
        if (application == nil)
        {
            UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Error initializing application"
                                                            message:error.description
                                                           delegate:self
                                                  cancelButtonTitle:@"OK"
                                                  otherButtonTitles:nil];
            [alert show];
            
            return;
        }
    }
    
    - (void)didReceiveMemoryWarning{
        [super didReceiveMemoryWarning];
    }
    
    /**
     This action will invoke the remove account APIs to clear the token cache
     to sign out a user from this application.
     */
    
    - (IBAction)signOut{
        
        MSALAccount *account = [self currentAccount];
        if (account == nil){
            return;
        }
        
        /**
         Removes all tokens from the cache for this application for the provided account
         
         - account:    The account to remove from the cache
         */
        
        [application removeAccount:account error:nil];
        _signout.enabled = false;
        [_resultTextView setText:@""];
    }
    
    // MARK: Acquiring and using token
    
    /**
     This will invoke the authorization flow.
     */
    
    - (IBAction)callGraphAction: (id)sender{
        
        MSALAccount *msalAccount = [self currentAccount];
        // We check to see if we have a current logged in account.
        // If we don't, then we need to sign someone in.
        if (msalAccount == nil){
            [self acquireTokenInteractively];
            return;
        }
        
        [self acquireTokenSilently:msalAccount];
    }
    
    -(void)acquireTokenSilently: (MSALAccount*) account {
        /**
         
         Acquire a token for an existing account silently
         
         - forScopes:           Permissions you want included in the access token received
         in the result in the completionBlock. Not all scopes are
         guaranteed to be included in the access token returned.
         - account:             An account object that we retrieved from the application object before that the
         authentication flow will be locked down to.
         - completionBlock:     The completion block that will be called when the authentication
         flow completes, or encounters an error.
         */
        NSArray *scopes = @[@"https://graph.microsoft.com/user.read"];
        
        MSALSilentTokenParameters *parameters = [[MSALSilentTokenParameters alloc] initWithScopes:scopes account:account];
        
        // interactionRequired means we need to ask the user to sign-in. This usually happens
        // when the user's Refresh Token is expired or if the user has changed their password
        // among other possible reasons.
        
        [application acquireTokenSilentWithParameters:parameters completionBlock:^(MSALResult *result, NSError *error)
         {
             if (error != nil){
                 if (error.domain == MSALErrorDomain) {
                     if (error.code == MSALErrorInteractionRequired){
                         dispatch_async(dispatch_get_main_queue(), ^{
                             [self acquireTokenInteractively];
                             return;
                         });
                     }
                 }
             }
             
             [self updateResultView:result];
             dispatch_async(dispatch_get_main_queue(), ^{
                 _signout.enabled = true ;
             });
             [self contentWithToken:[result accessToken]];
         }];
        
    }
    
    -(void)acquireTokenInteractively{
        NSArray *scopes = @[@"https://graph.microsoft.com/user.read"];
        MSALInteractiveTokenParameters *parameters = [[MSALInteractiveTokenParameters alloc] initWithScopes:scopes];
        
        
        void (^completionBlock)(MSALResult *result, NSError *error) = ^(MSALResult *result, NSError *error) {
            
            dispatch_async(dispatch_get_main_queue(), ^{
                
                if (result)
                {
                    [self updateResultView:result];
                    dispatch_async(dispatch_get_main_queue(), ^{
                        _signout.enabled = true ;
                    });
                    [self contentWithToken:[result accessToken]];
                }
                else
                {
                    [self updateResultViewError:error];
                }
                
            });
        };
        
        [application acquireTokenWithParameters:parameters completionBlock:completionBlock];
    }
    
    /**
     This will invoke the call to the Microsoft Graph API. It uses the
     built in URLSession to create a connection.
     */
    
    -(void) contentWithToken:(NSString *)token{
        
        // Specify the Graph API endpoint
        NSURL *graphURL = [NSURL URLWithString:(graphURI)];
        NSMutableURLRequest *graphRequest = [NSMutableURLRequest requestWithURL:graphURL];
        NSString *authValue = [NSString stringWithFormat:@"Bearer %@", token];
        // Set the Authorization header for the request. We use Bearer tokens, so we specify Bearer + the token we got from the result
        [graphRequest addValue:[NSString stringWithFormat:@"Bearer %@", token] forHTTPHeaderField:@"Authorization"];
        
        
        NSURLSession *session = [NSURLSession sharedSession];
        NSURLSessionDataTask *dataTask = [session dataTaskWithRequest:graphRequest
                                                    completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                                                        NSLog(@"%@",data);
                                                        if (error)
                                                        [self updateResultViewError:error];
                                                        else {
                                                            NSDictionary *json  = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
                                                            NSLog(@"%@",json);
                                                            dispatch_async(dispatch_get_main_queue(), ^{
                                                                [self.resultTextView setText:[json description]];
                                                            });
                                                        }
                                                    }];
        [dataTask resume];
        
    }
    
    -(void) updateResultView: (MSALResult*) result{
        NSString *resultText = [NSString stringWithFormat:@"{\n\taccessToken = %@\n\texpiresOn = %@\n\ttenantId = %@\n\tuser = %@\n\tscopes = %@\n\tauthority = %@\n}",
                                result.accessToken, result.expiresOn, result.tenantProfile.tenantId, result.account, result.scopes, result.authority];
        [_resultTextView setText:resultText];
        
    }
    
    -(void) updateResultViewError: (NSError*) error{
        NSString *resultErrorText = [NSString stringWithFormat:@"%@", error];
        [_resultTextView setText:resultErrorText];
    }
    
    // MARK: Get account and removing cache
    
    -(MSALAccount*) currentAccount{
        NSArray<MSALAccount *> *accounts = [application allAccounts:nil];
        // We retrieve our current account by getting the first account from cache
        // In multi-account applications, account should be retrieved by home account identifier or username instead
        return [accounts firstObject];
    }
    
    
    @end

File: MSALiOS/ViewController.swift
//------------------------------------------------------------------------------
//
// Copyright (c) Microsoft Corporation.
// All rights reserved.
//
// This code is licensed under the MIT License.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files(the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions :
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//
//------------------------------------------------------------------------------

import UIKit
import MSAL

/// 😃 A View Controller that will respond to the events of the Storyboard.

class ViewController: UIViewController, UITextFieldDelegate, URLSessionDelegate {
    
    let kIssuer = "https://login.microsoftonline.com/brandwedir.onmicrosoft.com/v2.0"
    let kClientID = "2a814505-ab4a-41f7-bd09-3fc614ac077c"
    let kRedirectURI = "msal://com.xerners/"
    let kLogoutURI = "https://login.microsoftonline.com/common/oauth2/v2.0/logout"
    let kGraphURI = "https://graph.microsoft.com/v1.0/me/"
    let kScopes: [String] = ["https://graph.microsoft.com/user.read"]
    let kAuthority = "https://login.microsoftonline.com/common/oauth2/v2.0/authorize"
    
    var accessToken = String()
    var applicationContext = MSALPublicClientApplication.init()
    

    @IBOutlet weak var loggingText: UITextView!
    @IBOutlet weak var signoutButton: UIButton!
    @IBOutlet weak var callGraphApiButton: UIButton!
    
    /**
     This button will invoke the authorization flow.
    */

@IBAction func authorizationButton(_ sender: UIButton) {
    
    
    do {
        
        // We check to see if we have a current logged in user. If we don't, then we need to sign someone in.
        // We throw an interactionRequired so that we trigger the interactive signin.
        
        
        if  try self.applicationContext.users().isEmpty {
            throw NSError.init(domain: "MSALErrorDomain", code: MSALErrorCode.interactionRequired.rawValue, userInfo: nil)
        } else {
        
        /**
         
         Acquire a token for an existing user silently
         
         - forScopes: Permissions you want included in the access token received
         in the result in the completionBlock. Not all scopes are
         gauranteed to be included in the access token returned.
         - User: A user object that we retrieved from the application object before that the
         authentication flow will be locked down to.
         - completionBlock: The completion block that will be called when the authentication
         flow completes, or encounters an error.
         */
        
        try self.applicationContext.acquireTokenSilent(forScopes: self.kScopes, user: applicationContext.users().first) { (result, error) in

                if error == nil {
                    self.accessToken = (result?.accessToken)!
                    self.loggingText.text = "Refreshing token silently)"
                    self.loggingText.text = "Refreshed Access token is \(self.accessToken)"
                    
                    self.signoutButton.isEnabled = true;
                    self.callGraphApiButton.isEnabled = true;

                } else {
                    self.loggingText.text = "Could not acquire token silently: \(error ?? "No error informarion" as! Error)"

                }
            }
        }
    }  catch let error as NSError {
        
        // interactionRequired means we need to ask the user to sign-in. This usually happens
        // when the user's Refresh Token is expired or if the user has changed their password
        // among other possible reasons.
        
        if error.code == MSALErrorCode.interactionRequired.rawValue {
            
            self.applicationContext.acquireToken(forScopes: self.kScopes) { (result, error) in
                    if error == nil {
                        self.accessToken = (result?.accessToken)!
                        self.loggingText.text = "Access token is \(self.accessToken)"
                        self.signoutButton.isEnabled = true;
                        self.callGraphApiButton.isEnabled = true;
                        
                    } else  {
                        self.loggingText.text = "Could not acquire token: \(error ?? "No error informarion" as! Error)"
                    }
            }
            
        }
        
    } catch {
        
        // This is the catch all error.
        
        self.loggingText.text = "Unable to acquire token. Got error: \(error)"
        
    }
    }

    
    /**
     This button will invoke the call to the Microsoft Graph API. It uses the
     built in Swift libraries to create a connection.
     Pay attention to the error case below. It shows you how to
     detect a `UserInteractionRequired` Error case and present the `acquireToken()`
     method again for the user to sign in. This usually happens if
     the Refresh Token has expired or the user has changed their
     password.
     
     */

@IBAction func callGraphApi(_ sender: UIButton) {
    
    let sessionConfig = URLSessionConfiguration.default
    
    // Specify the Graph API endpoint
    let url = URL(string: kGraphURI)
    var request = URLRequest(url: url!)
    
    // Set the Authorization header for the request. We use Bearer tokens, so we specify Bearer + the token we got from the result
    request.setValue("Bearer \(self.accessToken)", forHTTPHeaderField: "Authorization")
    let urlSession = URLSession(configuration: sessionConfig, delegate: self, delegateQueue: OperationQueue.main)
    
    urlSession.dataTask(with: request) { data, response, error in
        
        let result = try? JSONSerialization.jsonObject(with: data!, options: [])
                    if result != nil {
                
                self.loggingText.text = result.debugDescription
            }
        }.resume()
}
      /**
     This button will invoke the signout APIs to clear the token cache.
     
     */

@IBAction func signoutButton(_ sender: UIButton) {
    
        do {
            
            /**
             Removes all tokens from the cache for this application for the provided user
             
             - user:    The user to remove from the cache
             */
            
            try self.applicationContext.remove(self.applicationContext.users().first)
            self.signoutButton.isEnabled = false;
            self.callGraphApiButton.isEnabled = false;
            
        } catch let error {
            self.loggingText.text = "Received error signing user out: \(error)"
            }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        do {
            
            /**
             
             Initialize a MSALPublicClientApplication with a given clientID and authority
             
             - clientId:     The clientID of your application, you should get this from the app portal.
             - authority:    A URL indicating a directory that MSAL can use to obtain tokens. In Azure AD
             it is of the form https://<instance/<tenant>, where <instance> is the
             directory host (e.g. https://login.microsoftonline.com) and <tenant> is a
             identifier within the directory itself (e.g. a domain associated to the
             tenant, such as contoso.onmicrosoft.com, or the GUID representing the
             TenantID property of the directory)
             - Parameter error       The error that occurred creating the application object, if any, if you're
             not interested in the specific error pass in nil.
             
             */
            
            self.applicationContext = try MSALPublicClientApplication.init(clientId: kClientID, authority: kAuthority)
            
        } catch {
            
            self.loggingText.text = "Unable to create Application Context"
            
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func viewWillAppear(_ animated: Bool) {
        
        if self.accessToken.isEmpty {
            
            signoutButton.isEnabled = false;
            callGraphApiButton.isEnabled = false;
            
        }
    }

}

